glean-core/src/database/mod.rs

```diff
@@ -22,17 +22,28 @@ pub struct Database {
     // as the application lives: they don't need to be persisted
     // to disk using rkv. Store them in a map.
     app_lifetime_data: RwLock<BTreeMap<String, Metric>>,
+    // If the `defer_collection` Glean config option is `true`,
+    // we will save the 'ping' lifetime data in a map temporarily
+    // so as to persist them to disk using rkv in bulk on shutdown,
+    // or after a given interval, instead of everytime a new metric
+    // is created / updated.
+    ping_lifetime_data: Option<RwLock<BTreeMap<String, Metric>>>,
 }
 
 impl Database {
     /// Initialize the data store.
     ///
     /// This opens the underlying rkv store and creates
     /// the underlying directory structure.
-    pub fn new(data_path: &str) -> Result<Self> {
+    pub fn new(data_path: &str, defer_collection: bool) -> Result<Self> {
         Ok(Self {
             rkv: Self::open_rkv(data_path)?,
             app_lifetime_data: RwLock::new(BTreeMap::new()),
+            ping_lifetime_data: if defer_collection {
+                Some(RwLock::new(BTreeMap::new()))
```

Comment:
If we defer collection, we might still have `Lifetime::Ping` data around on startup.
We would need to load this data from disk in order for metrics to be valid (otherwise we would later override them when they are persisted again)

Reply:
This PR is only about [Bug 1596440](https://bugzilla.mozilla.org/show_bug.cgi?id=1596440) and @Dexterp37 advised me to do all the persisting when I work on [Bug 1596442](https://bugzilla.mozilla.org/show_bug.cgi?id=1596442), so, for now, there shouldn't ever be anything on disk.

I wonder if it would be easier to just work on both bugs in this PR. What do you think?

Reply:
I'm fine with splitting it out then, that does indeed make sense and makes the review easier.